### PR TITLE
[zeromq] Fix mingw build issues (24711)

### PR DESCRIPTION
**Pull request by k-128**
### What does your PR fix?
- <https://github.com/microsoft/vcpkg/issues/24711#issue-1235979108>

<br />

*Fix: zeromq build failure on mingw*
- Added `-DZMQ_HAVE_IPC=0` to `PLATFORM_OPTIONS`, 

<br />

*Fix: zeromq warning on builds*
- Set dynamic `LINKAGE_OPTIONS` flag list:
  - `WITH_PERF_TOOL` option only used if `-DBUILD_SHARED=ON`: 
    - <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1558>
  - option values are set explicitly to avoid issues with future libzmq changes

<br />

### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
- No changes were made to ci.baseline.txt

<br />

*Tested with*
```sh
# MSYS2: mingw64 env, toolchain v11.3.0
pacman -S --needed base-devel mingw-w64-x86_64-toolchain
./vcpkg.exe install zeromq:x86-mingw-static zeromq:x64-mingw-static zeromq:x64-mingw-dynamic zeromq:x86-windows zeromq:x64-windows zeromq:x64-windows-static

# MSYS2: mingw32 env, toolchain v11.3.0
pacman -S --needed base-devel mingw-w64-i686-toolchain
./vcpkg.exe install zeromq:x86-mingw-dynamic
# mingw32 required to avoid incorrect architecture with -DBUILD_SHARED=ON: "Expected x86, but was: x64"
```

<br />

### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
- yes

<br />

### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
- yes

<br />

### CODE REVIEW

File: ports/zeromq/portfile.cmake [hunk 1]
```diff
@@ -22,21 +19,25 @@ vcpkg_check_features(
 
 set(PLATFORM_OPTIONS)
 if(VCPKG_TARGET_IS_MINGW)
-    set(PLATFORM_OPTIONS "-DCMAKE_SYSTEM_VERSION=6.0")
+    set(PLATFORM_OPTIONS -DCMAKE_SYSTEM_VERSION=6.0 -DZMQ_HAVE_IPC=0)
```
Comment (JackBoosY) [inline ports/zeromq/portfile.cmake:22]: The only valid part of the whole modification is here? If yes, revert the modifications below.
We want to be able to see more clearly what the options are.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by k-128** [comment]
### Problem
---

The USE_PERF_TOOLS flag is only used, and should only be set if -DBUILD_SHARED=ON:
- <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1558>
- <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1569>
- <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1621>
  - `message(FATAL_ERROR "Shared library disabled - perf-tools unavailable.")`
  - Wrongly indented but clearly state the intent

The USE_PERF_TOOLS flag was always declared, triggering warnings on some builds, as I mentioned in <https://github.com/microsoft/vcpkg/issues/24711#issue-1235979108>

So I would disagree with the statement that the "The only valid part is..." (<https://github.com/microsoft/vcpkg/pull/24782#discussion_r876631378>), and all CI checks passed.

<br />

### Solution
---

Dynamically declare the USE_PERF_TOOLS flag if -DBUILD_SHARED=ON
- ON being the default *option* value: <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1304>
  - But as I said, to keep the port future proof these default option values shouldn't be used, as they are subject to changes.

<br />

### Current implementation
---

With the intent of saving space, seemingly a concern in other files and PRs:
- The USE_PERF_TOOLS flag is only declared with dynamic builds, instead of directly only when -DBUILD_SHARED=ON
- Single lines are used to declare LINKAGE_OPTIONS flags
- Comments aren't used, as it's almost never done in other ports, if deliberately to save space I will mention:
  - Better to lose a couple of lines in a file, than a couple of hours going through commits and PRs to understand choices

I'm not sure what is meant by "more clearly", I shared the relevant, libzmq CMakeLists lines, went into more details in the issue (<https://github.com/microsoft/vcpkg/issues/24711#issue-1235979108>) and this message, but I agree with long lines being indigest, redeclarations too, even dangerous.

For these reasons I would propose the following changes.

<br />

### Proposed changes
---

1: Both dynamically set and declared flags based on VCPKG_LIBRARY_LINKAGE

```cmake
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" BUILD_STATIC)
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" BUILD_SHARED)

set(LINKAGE_OPTIONS
    -DWITH_LIBSODIUM_STATIC=${BUILD_STATIC}
    -DBUILD_STATIC=${BUILD_STATIC}
    -DBUILD_SHARED=${BUILD_SHARED}
)
if(BUILD_SHARED)
    list(APPEND LINKAGE_OPTIONS -DWITH_PERF_TOOL=OFF)
endif()
```

<br />

2: With a comment on USE_PERF_TOOLS

```cmake
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" BUILD_STATIC)
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" BUILD_SHARED)

set(LINKAGE_OPTIONS
    -DWITH_LIBSODIUM_STATIC=${BUILD_STATIC}
    -DBUILD_STATIC=${BUILD_STATIC}
    -DBUILD_SHARED=${BUILD_SHARED}
)
if(BUILD_SHARED)
    # The USE_PERF_TOOLS option flag should only be set if -DBUILD_SHARED=ON:
    # - <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1621>
    list(APPEND LINKAGE_OPTIONS -DWITH_PERF_TOOL=OFF)
endif()
```

`LINKAGE_OPTIONS` could also be renamed.

<br />

Waiting for responses on the prefered ways to handle this.

**Review by JackBoosY — APPROVED** [review]

**Comment by k-128** [comment]
Thank you for your quick review.

Additional notes on the issues I fixed:
- The breaking changes were introduced the [Apr 30, 2021](<https://github.com/microsoft/vcpkg/commit/c8c95c50c11689920b9c67d88f9028f7553b389b>) or before, more than a year ago, with a [libzmq update](<https://github.com/zeromq/libzmq/issues/3691>)
- The warning issue was probably there for years.
- My PR was functionally valid from the start, it was trivial but challenged, declared invalid, for not following arbitrary undocumented styling rules, as a show of good faith I detailed my design choices as best I could, with an interest in learning more which wasn't met.

**Comment by dg0yt** [comment]
@k-128 Did you intentionally close the PR? It is not merged yet.

> * The breaking changes were introduced the [Apr 30, 2021](https://github.com/microsoft/vcpkg/commit/c8c95c50c11689920b9c67d88f9028f7553b389b) or before, more than a year ago, with a [libzmq update](https://github.com/zeromq/libzmq/issues/3691)

I take this as an indication that the vcpkg port doesn't  have users on mingw. Mingw support is mostly community-based here.

> * The warning issue was probably there for years.

The maintainer function which issues the warning learned this in the past 12 months, complemented by the `MAYBE_UNUSED_VARIABLES` option. These new warnings are usually adressed with other updates to the port,

> * My PR was functionally valid from the start, it was trivial but challenged, declared invalid, for not following arbitrary undocumented styling rules, as a show of good faith I detailed my design choices as best I could, with an interest in learning more which wasn't met.

I'm sure your PR is appreciated even if it the feedback didn't come as positively as you might have expected.
In fact I also thought hat initial PR seemed to change more than necessary for the things it tried to fix. This port has patterns which are also found in other ports, and retaining the patterns helps with maintenance. 

What I can add is that I would prefer to see the warning resolved with `MAYBE_UNUSED_VARIABLES` instead of putting `WITH_PERF_TOOL` into `LINKAGE_OPTIONS`.

**Comment by k-128** [comment]
Closed it thinking I was closing the issue my bad...

<br />

> I take this as an indication that the vcpkg port doesn't have users on mingw.

My thoughts too, or they fix their issues privately without contributing.

<br />

> What I can add is that I would prefer to see the warning resolved with MAYBE_UNUSED_VARIABLES instead of putting WITH_PERF_TOOL into LINKAGE_OPTIONS.

I mentioned this solution in: <https://github.com/microsoft/vcpkg/pull/24782#issuecomment-1131605514>
- "A less explicit solution could be to use vcpkg_cmake_configure MAYBE_UNUSED_VARIABLES to declare USE_PERF_TOOLS"

But a less obvious potential future issue could be the mistake I mentioned in libzmq CMakeLists.txt:
- <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1621>

```cmake
  if(WITH_PERF_TOOL)
    ...
  elseif(WITH_PERF_TOOL)
    message(FATAL_ERROR "Shared library disabled - perf-tools unavailable.")
  endif()
```

- The indentation is wrong, what was meant was:

```cmake
if(BUILD_SHARED)
  ...
elseif(WITH_PERF_TOOL)
  message(FATAL_ERROR "Shared library disabled - perf-tools unavailable.")
endif()
```

- If this gets fixed in libzmq, the declaration of `WITH_PERF_TOOL=ON` with `BUILD_SHARED=OFF` will trigger a FATAL_ERROR:
  - CMake Error, stop processing and generation, on builds (<https://cmake.org/cmake/help/latest/command/message.html>)

<br />

> I'm sure your PR is appreciated even if it the feedback didn't come as positively as you might have expected.
> In fact I also thought that the initial PR seemed to change more than necessary for the things it tried to fix. This port has patterns which are also found in other ports, and retaining the patterns helps with maintenance.

I was mainly looking for clarity, to get as many questions out of the way if I ever wanted to do another PR. I did actually try to inspire myself from multiple other ports, hence the mentioned note, raising the possibility of standardizing and documenting styling practices, I didn't like the look of my initial PR either, but it was functional.

Thank you for taking the time to go through all that.

**Comment by JackBoosY** [comment]
So is there any conclusion here?

**Comment by JackBoosY** [comment]
Ping for response.

**Comment by dg0yt** [comment]
@JackBoosY What is your opinion with regard to `MAYBE_UNUSED_VARIABLES`?

**Comment by JackBoosY** [comment]
> @JackBoosY What is your opinion with regard to `MAYBE_UNUSED_VARIABLES`?

We must verify the correctness of `MAYBE_UNUSED_VARIABLES` every time the port is changed, remove the expired items and add the items reported in the output.

**Comment by dg0yt** [comment]
> We must verify the correctness of `MAYBE_UNUSED_VARIABLES` every time the port is changed, remove the expired items and add the items reported in the output.

What does this mean with regard to this PR?

File: ports/zeromq/portfile.cmake [hunk 2]
```diff
@@ -20,23 +20,30 @@ vcpkg_check_features(
...
         -DBUILD_SHARED=${BUILD_SHARED}
-        -DWITH_PERF_TOOL=OFF
         -DWITH_DOCS=OFF
         -DWITH_NSS=OFF
         -DWITH_LIBSODIUM_STATIC=${BUILD_STATIC}
         ${FEATURE_OPTIONS}
         ${PLATFORM_OPTIONS}
+        ${LINKAGE_OPTIONS}
     OPTIONS_DEBUG
         "-DCMAKE_PDB_OUTPUT_DIRECTORY=${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg"
```
Comment (JackBoosY) [inline ports/zeromq/portfile.cmake:48]:
```suggestion
        "-DCMAKE_PDB_OUTPUT_DIRECTORY=${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg"
    MAYBE_UNUSED_VARIABLES
        USE_PERF_TOOLS
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please run `./vcpkg x-add-version --overwrite-version zeromq` then commit changes.

File: ports/zeromq/portfile.cmake [hunk 3]
```diff
@@ -20,25 +20,34 @@ vcpkg_check_features(
...
 if(VCPKG_TARGET_IS_MINGW)
-    set(PLATFORM_OPTIONS "-DCMAKE_SYSTEM_VERSION=6.0")
+    set(PLATFORM_OPTIONS -DCMAKE_SYSTEM_VERSION=6.0 -DZMQ_HAVE_IPC=0)
+endif()
+
+set(LINKAGE_OPTIONS "")
+if(BUILD_SHARED)
+    # The USE_PERF_TOOLS option flag is used if -DBUILD_SHARED=ON:
+    # - <https://github.com/zeromq/libzmq/blob/master/CMakeLists.txt#L1621>
+    list(APPEND LINKAGE_OPTIONS -DWITH_PERF_TOOL=OFF)
```
Comment (JackBoosY) [inline ports/zeromq/portfile.cmake:32]: I think it's not good.
The perf tool should be made as a feature and we should make some judgment here:
```cmake
if (BUILD_SHARED AND "tool" IN_LIST FEATURES)
    message(FATAL_ERROR "Perf tool should not be enabled when building dynamic.")
endif()
```
And `zeromq[*]` will be built fail when using dynamic triplet.
I think that's why the origin maintainer disabled this.

So can you please keep the origin code?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Pull request merged**